chardet/utf1632prober.py

```diff
@@ -0,0 +1,137 @@
+        if self.is_likely_utf16be():
+            return "utf-16be"
+        if self.is_likely_utf16le():
+            return "utf-16le"
+        # default to something valid
+        return "utf-16"
+
+    @property
+    def language(self):
+        return ""
+
+    def approx_32bit_chars(self):
+        return max(1.0, self.position / 4.0)
+
+    def approx_16bit_chars(self):
+        return max(1.0, self.position / 2.0)
+
+    def is_likely_utf32be(self):
+        approx_chars = self.approx_32bit_chars()
+        return approx_chars >= self.MIN_CHARS_FOR_DETECTION and (
+            self.zeros_at_mod[0] / approx_chars > self.EXPECTED_RATIO and
+            self.zeros_at_mod[1] / approx_chars > self.EXPECTED_RATIO and
+            self.zeros_at_mod[2] / approx_chars > self.EXPECTED_RATIO and
+            self.nonzeros_at_mod[3] / approx_chars > self.EXPECTED_RATIO)
+
+    def is_likely_utf32le(self):
+        approx_chars = self.approx_32bit_chars()
+        return approx_chars >= self.MIN_CHARS_FOR_DETECTION and (
+            self.nonzeros_at_mod[0] / approx_chars > self.EXPECTED_RATIO and
+            self.zeros_at_mod[1] / approx_chars > self.EXPECTED_RATIO and
+            self.zeros_at_mod[2] / approx_chars > self.EXPECTED_RATIO and
+            self.zeros_at_mod[3] / approx_chars > self.EXPECTED_RATIO)
+
+    def is_likely_utf16be(self):
+        approx_chars = self.approx_16bit_chars()
+        return approx_chars >= self.MIN_CHARS_FOR_DETECTION and (
+            (self.nonzeros_at_mod[1] + self.nonzeros_at_mod[3]) / approx_chars > self.EXPECTED_RATIO and
+            (self.zeros_at_mod[0] + self.zeros_at_mod[2]) / approx_chars > self.EXPECTED_RATIO)
+
+    def is_likely_utf16le(self):
+        approx_chars = self.approx_16bit_chars()
+        return approx_chars >= self.MIN_CHARS_FOR_DETECTION and (
+            (self.nonzeros_at_mod[0] + self.nonzeros_at_mod[2]) / approx_chars > self.EXPECTED_RATIO and
+            (self.zeros_at_mod[1] + self.zeros_at_mod[3]) / approx_chars > self.EXPECTED_RATIO)
+
+    def feed(self, byte_str):
+        for c in byte_str:
+            if c == 0:
+                self.zeros_at_mod[self.position % 4] += 1
+            else:
+                self.nonzeros_at_mod[self.position % 4] += 1
+            self.position += 1
+        return self.state()
+
+    def state(self):
+        if self._state in [ProbingState.NOT_ME, ProbingState.FOUND_IT]:
+            # terminal, decided states
+            return self._state
+        elif self.get_confidence() > 0.95:
+            self._state = ProbingState.FOUND_IT
+        elif self.position > 4 * 1024:
+            # if we get to 4kb into the file, and we can't conclude it's UTF,
+            # let's give up
+            self._state = ProbingState.NOT_ME
+        return self._state
+
+    def get_confidence(self):
+        confidence = 0.99
+
+        if self.is_likely_utf16le() or self.is_likely_utf16be() or self.is_likely_utf32le() or self.is_likely_utf32be():
```

Comment:
I'd really like to see a more mathematical basis for the confidence here, rather than the essentially binary 0.99 or 0.01 we have now. Maybe something based on the difference between `EXPECTED_RATIO` and the values you have in `nonzeros_at_mod` and `zeros_at_mod`? 0.99 is a really high confidence, and I like to reserve it for definite yeses. In fact, calculating this confidence is a lot of what the `CharDistributionAnalysis` class does, which is part of why I think that's a more natural fit for this codec.